internal/database/database.go

```diff
@@ -0,0 +1,485 @@
+// Package database provides persistent storage for auth tokens, jobs, and flavors,
+// and computes "pressure" for each flavor.
+//
+// Pressure is a numeric measure of demand for runner machines (flavors). It is
+// defined as the number of incomplete jobs (completed_at IS NULL) assigned to a
+// given flavor.
+//
+// When a job is inserted, it is assigned a flavor using the following algorithm:
+//  1. The flavor's platform must match the job's platform.
+//  2. The flavor's labels must be a superset of (or equal to) the job's labels.
+//  3. If multiple flavors match, choose the one with the highest priority.
+//  4. If there is still a tie, pick one at random.
+package database
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+const (
+	DefaultLimit            = 100
+	updateAssignedFlavorSql = `
+	UPDATE job AS j
+	SET assigned_flavor = (SELECT f.name
+						   FROM flavor AS f
+						   WHERE f.is_disabled = FALSE
+							 AND f.platform = j.platform
+							 AND f.labels @> j.labels
+						   ORDER BY f.priority DESC, random()
+						   LIMIT 1)
+	WHERE j.assigned_flavor IS NULL AND j.completed_at IS NULL
+	  AND EXISTS (SELECT 1
+				  FROM flavor AS f
+				  WHERE f.is_disabled = FALSE
+					AND f.platform = j.platform
+					AND f.labels @> j.labels
+				  );
+`
+)
+
+var (
+	ErrNotExist = errors.New("does not exist")
+	ErrExist    = errors.New("already exists")
+)
+
+type Database struct {
+	conn *pgxpool.Pool
+}
+
+type ListJobOptions struct {
+	WithId           string
+	OnlyNotCompleted bool
+	CreatedAfter     time.Time
+	Limit            int
+}
+
+// CreateAuthToken creates an authentication token for the given name.
```

Comment:
Question (convention): Is it Go convention to put comments above using single line comments?

Reply:
This is the equivalent of Python docstrings, called [go doc](https://tip.golang.org/doc/comment) for Go functions. It needs to be at the beginning of a function.